src/ucp/wireup/wireup_cm.c

```diff
@@ -1068,8 +1095,33 @@ void ucp_cm_server_conn_request_cb(uct_listener_h listener, void *arg,
                      UCT_DEVICE_NAME_MAX);
     memcpy(ucp_conn_request->remote_dev_addr, remote_data->dev_addr,
            remote_data->dev_addr_length);
+
+    /* [ sa_data | ucp_address | user_data ]
+     * |      conn_priv_data_length        |
+     * 1. sa_data is fixed of size, so copy it directly */
     memcpy(&ucp_conn_request->sa_data, remote_data->conn_priv_data,
-           remote_data->conn_priv_data_length);
+           sizeof(ucp_conn_request->sa_data));
+
+    /* 2. Unpack ucp_address and get it's length in buffer */
+    addr_flags                    = ucp_cm_address_pack_flags(worker);
+    ucp_conn_request->remote_addr = ucs_malloc(sizeof(ucp_unpacked_address_t),
+                                               "ucp_unpacked_address_t");
+
+    status = ucp_address_unpack(ucp_conn_request->listener->worker,
+                                UCS_PTR_BYTE_OFFSET(remote_data->conn_priv_data,
+                                                    sizeof(ucp_conn_request->sa_data)),
+                                addr_flags, ucp_conn_request->remote_addr,
+                                &ucp_addr_size);
+    if (status != UCS_OK) {
+        goto err_free_remote_addr;
+    }
+
+    /* 3. Copy user_data */
+    memcpy(&ucp_conn_request->user_data,
```

Comment:
is it possible user_data is not present?

Reply:
Yes, it's possible and if user will query ucp_conn_request will get invalid data. If user didn't set client_id but query it - it's invalid usage. Or we can add some additional address mode flag, and check it here.

Reply:
so it means we could access invalid memory here? need to avoid it

Reply:
yes, it breaks compatibility

Reply:
Added check, copy user_data only if `priv_data_length - sa_data - unpacked_address_size == sizeof(user_data)`